src/hither/createHitherJob.js

```diff
@@ -1,216 +1,210 @@
 import File from './File'
 import { addHitherJob, updateHitherJob } from '../actions/hitherJobs'
 
-const axios = require('axios');
 const objectHash = require('object-hash');
 
-const globalHitherJobStore = {};
-const globalDispatch = {};
-let globalJobHandlersByRole = {};
+const globalData = {
+  dispatch: null,
+  apiConnection: null,
+  hitherClientJobCache: {},
+  hitherJobs: {},
+  runningJobIds: {}
+};
 
 export const sleepMsec = m => new Promise(r => setTimeout(r, m));
 
+export const handleHitherJobCreated = (msg) => {
+  if (!(msg.client_job_id in globalData.hitherJobs)) {
+    console.warn(`Unable to find job with clientJobId: ${msg.client_job_id}.`);
+    return;
+  }
+  const job = globalData.hitherJobs[msg.client_job_id];
+  job._handleHitherJobCreated({jobId: msg.job_id});
+  globalData.hitherJobs[msg.job_id] = job;
+  delete globalData.hitherJobs[job.clientJobId];
+  dispatchAddHitherJob(job.object());
+}
+
+export const handleHitherJobCreationError = (msg) => {
+  if (!(msg.client_job_id in globalData.hitherJobs)) {
+    console.warn('Unable to find job with clientJobId (hitherJobCreationError).');
+    return;
+  }
+  const job = globalData.hitherJobs[msg.client_job_id];
+  const errorString = `Error creating job ${job._functionName}: ${msg.error}`;
+  job._handleHitherJobCreationError({
+    errorString,
+    runtime_info: null
+  });
+  delete globalData.hitherJobs[job.clientJobId];
+}
+
+export const handleHitherJobFinished = (msg) => {
+  const job = globalData.hitherJobs[msg.job_id] || globalData.hitherJobs[msg.client_job_id];
+  if (!job) {
+    console.warn(`job not found (handleHitherJobFinished): ${msg.job_id} ${msg.client_job_id}`);
+    return;
+  }
+  job._handleHitherJobFinished({
+    result: msg.result,
+    runtime_info: msg.runtime_info
+  })
+  dispatchUpdateHitherJob({jobId: job._jobId, update: job.object()});
+}
+
+export const handleHitherJobError = (msg) => {
+  const job = globalData.hitherJobs[msg.job_id] || globalData.hitherJobs[msg.client_job_id];
+  if (!job) {
+    console.warn(`job not found (handleHitherJobError): ${msg.job_id} ${msg.client_job_id}`);
+    return;
+  }
+  job._handleHitherJobError({
+    errorString: msg.error_message,
+    runtime_info: msg.runtime_info
+  })
+  dispatchUpdateHitherJob({jobId: job._jobId, update: job.object()});
+}
+
 const dispatchAddHitherJob = (job) => {
-  if (!globalDispatch.dispatch) return;
-  globalDispatch.dispatch(addHitherJob(job));
+  if (!globalData.dispatch) return;
```

Comment:
I reckon this is probably a "can't-happen" sort of thing, but should it throw an error if it does occur, or is it expected?